src/FsAutoComplete/CodeFixes/IgnoreExpression.fs

```diff
@@ -0,0 +1,104 @@
+
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.SyntaxTrivia
+open FSharp.Compiler.Text
+open FSharp.UMX
+open FsToolkit.ErrorHandling
+open Ionide.LanguageServerProtocol.Types
+open FsAutoComplete.CodeFix.Types
+open FsAutoComplete
+open FsAutoComplete.LspHelpers
+
+let title = "Ignore expression"
+
+let fix (getParseResultsForFile: GetParseResultsForFile) : CodeFix =
+  Run.ifDiagnosticByCode (set [ "20" ]) (fun diagnostic (codeActionParams: CodeActionParams) ->
+    asyncResult {
+      let fileName = codeActionParams.TextDocument.GetFilePath() |> Utils.normalizePath
+      let fcsPos = protocolPosToPos codeActionParams.Range.Start
+      let mDiag = protocolRangeToRange (UMX.untag fileName) diagnostic.Range
+
+      // Only do single line for now
+      if mDiag.StartLine <> mDiag.EndLine then
+        return []
+      else
+
+      let! (parseAndCheckResults: ParseAndCheckResults, _line: string, sourceText: IFSACSourceText) =
+        getParseResultsForFile fileName fcsPos
+
+      let mExprOpt =
+        (fcsPos, parseAndCheckResults.GetParseResults.ParseTree)
+        ||> ParsedInput.tryPick (fun path node ->
+          match node with
+          | SyntaxNode.SynExpr(e) when Range.equals mDiag e.Range -> Some(path, e)
+          | _ -> None)
+
+      match mExprOpt with
+      | None -> return []
+      | Some(path, expr) ->
+
+      let needsParentheses =
+        let appPipe, appIgnore =
+          let lineNumber = expr.Range.StartLine
+
+          let mkRangeInFile (offset: int) (length: int) : range =
+            Range.mkRange
+              expr.Range.FileName
+              (Position.mkPos lineNumber (expr.Range.EndColumn + offset))
+              (Position.mkPos lineNumber (expr.Range.EndColumn + offset + length))
+
+          let mPipe = mkRangeInFile 2 2
+          let mIgnore = mkRangeInFile 5 6
+
+          let appPipe =
+            SynExpr.App(
+              ExprAtomicFlag.NonAtomic,
+              true,
+              SynExpr.LongIdent(
+                false,
+                SynLongIdent(
+                  [ FSharp.Compiler.Syntax.Ident("op_PipeRight", mPipe) ],
+                  [],
+                  [ Some(IdentTrivia.OriginalNotation "|>") ]
+                ),
+                None,
+                mPipe
+              ),
+              expr, // The expr that will now be piped into ignore.
+              Range.unionRanges expr.Range mPipe
+            )
+
+          let appIgnore =
+            SynExpr.App(
+              ExprAtomicFlag.NonAtomic,
+              false,
+              appPipe,
+              SynExpr.Ident(FSharp.Compiler.Syntax.Ident("ignore", mIgnore)),
+              Range.unionRanges expr.Range mIgnore
+            )
+
+          appPipe, appIgnore
+
+        let pathWithPipeIgnore =
+          SyntaxNode.SynExpr appPipe :: SyntaxNode.SynExpr appIgnore :: path
+
+        SynExpr.shouldBeParenthesizedInContext sourceText.GetLineString pathWithPipeIgnore expr
```

Comment:
Hmm, this looks right — but it seems that `SynExpr.shouldBeParenthesizedInContext` still doesn't behave the way one might expect. I believe it's because of this:

https://github.com/dotnet/fsharp/blob/c4905b18d48caf52b105a3bbbca57c93c600c60b/src/Compiler/Service/SynExpr.fs#L504-L505

```fsharp
    let containsSensitiveIndentation (getSourceLineStr: int -> string) outerOffsidesColumn (range: range) =
        let startLine = range.StartLine
        let endLine = range.EndLine

        if startLine = endLine then
            range.StartColumn <= outerOffsidesColumn
        else
            // …
```

The above is causing `containsSensitiveIndentation` to [return `true`](https://github.com/dotnet/fsharp/blob/c4905b18d48caf52b105a3bbbca57c93c600c60b/src/Compiler/Service/SynExpr.fs#L655-L660) in every test case here, since the offsides line is the start of the line, and since of course there were *not* originally any parentheses around the expression, the expression's range starts at the offsides column. That *would* represent "sensitive indentation" if the expression were already parenthesized, but it doesn't when it isn't.

The common case when this API is used for the parens analyzer looks like this:

```fsharp
// This is the outer offsides column.
//  ↓
    (9) |> ignore
//   ↑
// The inner expr's start column is > the outer offsides.
```

Since, when we're checking whether parens are needed in such a case, we're _really_ checking whether we can _remove_ parens that are already there, the inner expr's start column for a single-line parenthesized expression will never be the same as the outer offsides column, since the inner expr's start column will always be shifted at least one column to the right by the open paren.

The hypothetical that we're giving it here is instead:

```fsharp
// This is the outer offsides column.
//  ↓
    9 |> ignore
//  ↑
// The inner expr's start column = the outer offsides.
```

Since `range.StartColumn <= outerOffsidesColumn`, `containsSensitiveIndentation` returns `true`.

The reason `containsSensitiveIndentation` has this test is to handle situations like this:

```fsharp
async {
// This is the outer offsides column.
// The inner expr's start column = the outer offsides.
//  ↓
    return
     (
    9
     )
}
```

or this:

```fsharp
async {
// This is the outer offsides column.
// The inner expr's start column = the outer offsides.
//  ↓
    return (
    9
    )
}
```

which compile with the parentheses, but which would be invalid without them.

Maybe I can update it to be a bit smarter and only return `true` when the outer construct is a construct known to allow such sensitive indentation to begin with? (Like `return`, etc.)

Reply:
Thanks, I hope we find this problem in time for the 8.0.300 release.
A bit fearful for that though.

Reply:
It would be best if a good could be identified and merged next week if that's the case - after that it gets much harder to get changes into 8.0.300.

Reply:
I think this PR should address this immediate problem: https://github.com/dotnet/fsharp/pull/16973

Reply:
https://github.com/dotnet/fsharp/pull/16973 is now part of this PR.